sycl/source/device_selector.cpp

```diff
@@ -15,76 +15,133 @@
 
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
+
+// Utility function to check if device is of the preferred SYCL_BE.
+static bool isDeviceOfPreferredSyclBe(const device &Device) {
+  // Taking the version information from the platform gives us more useful
+  // information than the driver_version of the device.
+  const platform Platform = Device.get_info<info::device::platform>();
+  const std::string PlatformVersion =
+      Platform.get_info<info::platform::version>();
+
+  detail::pi::Backend PreferredBE = detail::pi::getPreferredBE();
+
+  if (PreferredBE == RT::Backend::SYCL_BE_PI_OPENCL) {
+    if (PlatformVersion.find("OpenCL") != std::string::npos) {
+      return true;
+    }
+  } else if (PreferredBE == RT::Backend::SYCL_BE_PI_CUDA) {
+    if (PlatformVersion.find("CUDA") != std::string::npos) {
```

Comment:
The PlatformVersion value for NVIDIA OpenCL also contains the string CUDA, so this will pick up OpenCL NVIDIA first. See output of NVIDIA OpenCL platform from clinfo:

```
Patform Vendor                                 NVIDIA Corporation
Platform Version                                OpenCL 1.2 CUDA 10.1.236     
```

The solution is to change the string of the PI CUDA backend to something more specific that we can search or alternatively find CUDA string but ensure the OpenCL one is not there.

Reply:
actually the `useBackend` or now `preferredBackend` isn't used anywhere else than the device_selector with this change. Do we really need that runtime code in `pi.hpp`, or will it be simpler to evaluate the ENV in device selector? especially as `preferredBackend` will always return `OpenCl` when no env is set :)

Reply:
I now think we should just get the BE from the device's plugin, instead of querying their names. 
~~~
auto BE = detail::getSyclObjImpl(dev)->getPlugin().getBackend();
~~~
Then the `isDeviceOfPreferredSyclBe` can be removed.

Reply:
Fixed.